### DISCUSSION TITLE
Deploying / installing libvirt

### DISCUSSION THREAD

**Discussion by Tenebresus** [siderolabs/talos#9136] [Q&A]
Hi all,

Im currently trying to deploy OpenStack on my Talos cluster using the folllowing guide https://docs.openstack.org/openstack-helm/latest/install/openstack.html. For OpenStack to work on Kubernetes I have to deploy libvirt pods, whose jobs it is to start libvirt on the worker nodes (https://github.com/openstack/openstack-helm-infra/blob/master/libvirt/templates/bin/_libvirt.sh.tpl). This script is trying to call `systemd-run libvirtd` in order to start libvirt, but because talos uses it own init system (machined), this fails. Does anyone have any experience with deploying / installing libvirt on their talos cluster? If so, could you be so kind to help me out? :)

**Comment by smira** [comment]
Sorry, no experience, so just a generic answer.

Talos doesn't have `systemd`, so there should be some equivalent way to run `libvirtd`, I guess it might be a static pod, or a daemonset, or something like that.

**Reply by Tenebresus** [reply]
Do you have any experience with the init system that Talos uses? I am a bit hesitant with just deploying `libvirtd` in its own static pod, because another OpenStack service (nova) depends on that `libvirtd` is running on the host node and that it can connect to its running socket at `/run/libvirt/libvirt-sock` (https://github.com/openstack/openstack-helm/blob/5b6312f6207490fa5b72527726fe2a3b94aeb1c5/nova/values.yaml#L1454C51-L1454C76). Further I am not that familiar with Talos and I am still learning, so some concepts are still unfamiliar with me.

**Reply by smira** [reply]
Talos Linux is an immutable operating system, so many concepts don't apply directly.

OpenStack was not designed to run on Kubernetes, it was an afterthought I guess. In the end via host filesystem mounts you can achieve pretty similar result no matter how you run it. 

KubeVirt is much more Kubernetes-native "OpenStack".

**Reply by Tenebresus** [reply]
Ah right! What do you mean exactly with `KubeVirt` being a more Kubernetes-native "OpenStack"? Meaning that `KubeVirt` should replace `libvirt` or `OpenStack` itself? And could you also explain a bit more about "In the end via host filesystem mounts you can achieve pretty similar result no matter how you run it"? If you mean that i can use `KubeVirt` instead of `libvirt` how can i achieve this and that `nova` can still work and reference `KubeVirt` instead of `libvirt`?

**Reply by smira** [reply]
I mean if your goal is to run VMs on top of Kubernetes, KubeVirt provides a way more native Kubernetes experience.

If your goal is to run OpenStack, then you need to git further into OpenStack and ignore my KubeVirt suggestion.

**Reply by Tenebresus** [reply]
My main goal is to setup OpenStack and I choose Kubernetes, because containerization is the way to go. Talos advertising that its a super secure Kubernetes solution made it really appealing. I hope that its possible to launch VMs some way on Talos and also using that approach with Nova. I really appreciate your input on this; do you have any last thoughts or tips for how i can go further?

**Reply by smira** [reply]
I know that some users run OpenStack on Talos, but you might need to rework the instructions to run everything in containers, including `libvirt` components. I don't have anything specific sorry

**Reply by Tenebresus** [reply]
By any chance you have some references or know someone who has done this before?

**Reply by smira** [reply]
no

**Reply by steverfrancis** [reply]
I know some - I'll ask them if they can comment here with how they got things up

**Reply by Tenebresus** [reply]
That would be awesome, thanks!

**Reply by wvh-github** [reply]
We are familiar with the problem and the script. The concept behind it is simple. It starts an instance of libvirt outside of the container. It achieves this by utilizing a systemd feature to create a unit on the fly. This way the image, and with it libvirt +qemu, get distributed using the mechanics of kubernetes, but libvirt created vms are not running inside of the container. We have worked around this by letting libvirt start inside the container for now. We are in the midst of designing a good architecture for us to achieve a similar outcome to the original intent of running libvirt outside of the container.
@Tenebresus what is your current setup and what is the scale of the deployment? I'm interested in your goals with this.

**Reply by Tenebresus** [reply]
We currently have a POC Talos cluster running with 3 controller nodes and 3 worker nodes. I am trying to deploy OpenStack (version 2023.2) with the helm charts described in the OpenStack Helm documentation (https://docs.openstack.org/openstack-helm/latest/install/openstack.html). Thus far i have deployed everything (with also some work-arounds, because of Talos) up until libvirt and nova. I am really interested in the way you worked around this and how you deployed libvirt inside the container. If you need / want more information, let me know!

P.S. I see in your bio that you work for 42on, which is funny because my company and I are currently talking to your partners over at Fairbanks :)

**Reply by wvh-github** [reply]
Ah, that is interesting. I'm indeed with 42on and Fairbanks, we are part of the same parent company and work as close together as is possible. We sit in the same office. Are you talking to us through the other channel regarding this, or something else?
Maybe that is a more direct way of engaging.

**Reply by Tenebresus** [reply]
Im currently mailing with someone from your end to organize a meeting for consultancy, we've also got help from you guys with our ceph setup and were really happy with the help. But for now we want to gain as much knowledge about OpenStack on Talos and get something running before we tear it all down and build something production ready with the consultants. Could we maybe discuss the libvirt solution somewhere more direct?

**Reply by wvh-github** [reply]
Which timezone are you in?

**Reply by Tenebresus** [reply]
I am also from the Netherlands :)

**Reply by wvh-github** [reply]
Are you available monday?

**Reply by Tenebresus** [reply]
Yes, i am available on monday, would you like to schedule a meeting? is their a way that i can contact you?

**Reply by Tenebresus** [reply]
Hi @wvh-github, are you still available today? :)

**Reply by nitrocb** [reply]
Hi @Tenebresus @wvh-github,

I have a similar usecase here. I am trying to deploy [Ironcore](https://github.com/ironcore-dev), which is a k8s driven IaaS Solution utilizing libvirt to build VMs on Kubernetes Nodes. We want to use it on Talos, because of ease of setup and security reasons.

The Ironcore solution does not implement libvirt itself, so I needed to build a container image including libvirt to run it as a DaemonSet, which makes the libvirt socket available on the Worker nodes to be used by Ironcore.

The problem I run into is, when I start my image including libvirt on Talos and try to run virtlogd and libvirtd, the shell just hangs and nothing happens. 
When I run the container image on an ubuntu host, everything works fine and I can start VMs.

As you seem to have a bit of experience with libvirt on Talos linux, can you give me a hint, if there are some pre-requisites on Talos Linux, which I need to configure to be able to run libvirt on it?

Thanks a lot in advance!

Best regards

**Reply by wvh-github** [reply]
Hi @nitrocb ,
We are currently working with @Tenebresus on their specific use case through other means.

I'm sure I can help, but I'm not yet familiar with IronCore. If Ironcore expects libvirt to run on the host, there is no real way of doing that at the moment. I also think there is no good way of doing that, and you shouldn't need to really.

Are you per chance sharing any filesystem locations with the host etc? Is the DaemonSet running priviliged? How about the PID space, is that host based or not?

**Reply by nitrocb** [reply]
Hi @wvh-github ,

thanks for your answer!

Iconcore does only expect to have the libvirt socket available somewhere, where it can access it. So the idea would be to run libvirt in a pod and create the socket on a hostpath mount, where it can also be accessed by the ironcore pod.

I currently only share `/run/dbus/system_bus_socket` and `/dev/kvm` with the host. But I also removed those mounts just to check if this causes the problem.

The DaemonSet is running privileged and is using the host network.

hostPID is not set to true, so I guess the default will be false. I could try setting that to true.

I can not start virtlogd and libvirtd as they only hang and do not output or log anything.
I already created an strace, which I am currently analyzing.

**Reply by sjabasti** [reply]
Hi @wvh-github @Tenebresus 

We are in the exact same situation, openstack-helm on talos and libvirtd write mounting /etc/libvirt and the script requiring systemd-run. How did you solve these problems/running libvirtd inside a container?

Thanks in advance

**Comment by wvh-github** [comment]
I'm going to get back to you on this. Sharing dbus will confuse libvirt as it has hardcoded things it will try to do that might not actually work. We are not sharing those.
I'm going to check how I can share something workable for you.

**Reply by nitrocb** [reply]
Thanks a lot, really appreciate it!

**Reply by nitrocb** [reply]
Hi again @wvh-github,

your hint with the dbus socket mount from the host was indeed helpful.
I removed the hostpath mount and startet a dbus-daemon inside the container, which brought libvirt to work in the end.
I was already able to create a test-VM 😊
I am stil struggeling with the network part (Bridged networking, ipv6 vlan, custom mtu), but I'll figure that out.

Thanks a lot!

Best regards

**Reply by wvh-github** [reply]
Thanks for the feedback @nitrocb.
